File: viper/src/main/java/palestra/viper/router/AbstractRouter.java
package palestra.viper.router;

import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;

import palestra.viper.view.AndroidComponent;

public abstract class AbstractRouter<L extends MvpRouter.Listener> implements MvpRouter<L> {
    private L listener;

    @Override
    public void setListener(L listener) {
        this.listener = listener;
    }

    protected L getListener() {
        return listener;
    }

    protected void displayFragmentAdd(AndroidComponent androidComponent, Fragment fragment, String tag, int fragmentContainerId) {
        final FragmentManager fragmentManager = androidComponent.getSupportFragmentManager();
        if (fragmentManager.findFragmentByTag(tag) == null) {
            fragmentManager.beginTransaction()
                    .add(fragmentContainerId, fragment, tag)
                    .addToBackStack(null)
                    .commit();
        }
    }

    protected void displayFragmentReplace(AndroidComponent androidComponent, Fragment fragment, String tag, int fragmentContainerId) {
        final FragmentManager fragmentManager = androidComponent.getSupportFragmentManager();
        if (fragmentManager.findFragmentByTag(tag) == null) {
            fragmentManager.beginTransaction()
                    .replace(fragmentContainerId, fragment, tag)
                    .commit();
        }
    }
}

File: app/src/main/java/ru/pavlov/palestra/presentation/arch/AbstractRouter.kt
package ru.pavlov.palestra.presentation.arch

import android.content.Intent
import android.support.v4.app.ActivityOptionsCompat
import android.support.v4.app.Fragment
import android.view.View
import com.vershininds.mixture.router.AbstractRouter
import com.vershininds.mixture.router.MvpRouter
import com.vershininds.mixture.view.AndroidComponent

abstract class AbstractRouter<L : MvpRouter.Listener> : AbstractRouter<L>() {

    protected fun displayFragmentAdd(androidComponent: AndroidComponent, fragment: Fragment, tag: String, fragmentContainerId: Int) {
        val fragmentManager = androidComponent.supportFragmentManager
        if (fragmentManager.findFragmentByTag(tag) == null) {
            fragmentManager.beginTransaction()
                    .add(fragmentContainerId, fragment, tag)
                    .addToBackStack(null)
                    .commit()
        }
    }

    protected fun displayFragmentReplace(androidComponent: AndroidComponent, fragment: Fragment, tag: String, fragmentContainerId: Int) {
        val fragmentManager = androidComponent.supportFragmentManager
        if (fragmentManager.findFragmentByTag(tag) == null) {
            fragmentManager.beginTransaction()
                    .replace(fragmentContainerId, fragment, tag)
                    .commit()
        }
    }

    protected fun displayActivityWithSharedView(androidComponent: AndroidComponent, intent: Intent, sharedView: View, sharedViewName: String) {
        val activity = androidComponent.activity
        val activityOptionsCompat =
                ActivityOptionsCompat.makeSceneTransitionAnimation(activity, sharedView, sharedViewName).toBundle()

        intent.apply {
            activity.startActivity(this, activityOptionsCompat)
        }
    }
}